Replace specific potion/scroll generation rules with a generic system.

flend opened the pull request:
RapidBrogue extends these two item generation mechanisms from the base game (frequency bias and thresholding) to other items. This patch extends the item-specific code in the base game to allow these mechanisms to be used for any potion or scroll with parameters in Global.c. Tested against a few seeds to ensure identical generation.

Nathan-Fenner commented on src/brogue/Globals.c:
```diff
@@ -2406,6 +2406,43 @@ const char armorRunicNames[NUMBER_ARMOR_ENCHANT_KINDS][30] = {
+// incrementFreq must be != 0 for frequency biasing
+// lvlScale != 0 for thresholding
+//  { category, kind,                   initFreq,   incFreq,decFreq,gMulti, gInc,   lvlScale }
+const meteredItemGenerationTable meteredItemsGenerationTable[NUMBER_METERED_ITEMS] = {
```
Since this is a brand new table, I think it could make sense to make use of designated initializers here:

```c
const meteredItemGenerationTable meteredItemsGenerationTable[NUMBER_METERED_ITEMS] = {
  { .category = SCROLL, .kind = SCROLL_ENCHANTING, .initialFrequency = 60, .incrementFrequency = 30, .decrementFrequency = 50 },
  { .category = SCROLL, .kind = SCROLL_IDENTIFY },
  { .category = SCROLL, .kind = SCROLL_TELEPORT },
  { .category = SCROLL, .kind = SCROLL_REMOVE_CURSE },
  // ...
  { .category = POTION, .kind = POTION_LIFE, .initialFrequency = 0, .incrementFrequency = 34, .decrementFrequency = 150, .genMultiplier = 4, .genIncrement = 3, .levelScaling = 1},
  // ...
};
```

Although it makes the table a little bit more verbose, it means that the comment explaining the struct field order is not needed, and it's now guaranteed to be kept in-sync. Also, any zero fields (which is most of them) can be omitted entirely.

For easy-of-scanning, aligning them with spaces still makes sense, it will just be wider than it is now.

flend replied:
I like this change, view @tmewett ?

tmewett replied:
I don't mind!

tmewett reviewed (changes requested):
This is a good idea and I think the overall approach is pretty good.

Not blocking, but I would like to improve the table. Not keen on having to enumerate all scrolls and potions (in that order) due to a certain loop, and hence the use of `incrementFrequency != 0` as a gate for the behaviour. But happy to merge that as-is and adjust later?

flend commented:
I have addressed the else formatting and rewritten the table with designated identifiers. I have rewritten the table to avoid iteration (using a lookup etc.?) - to be discussed. Hopefully non blocking still, as above.

tmewett commented:
Hmm, I can't see the diff from the latest push for some reason. In any case, current code is good for current purposes.

Question though, why do we save and restore the tables? Are the un-overridden values used in other places?

flend commented:
> Question though, why do we save and restore the tables? Are the un-overridden values used in other places?

This replaces the original 

```scrollTable[SCROLL_ENCHANTING].frequency    = 0;    // No enchant scrolls or strength/life potions can spawn except via initial
    potionTable[POTION_STRENGTH].frequency      = 0;    // item population or blueprints that create them specifically.
    potionTable[POTION_LIFE].frequency          = 0;```

So I was just matching current behaviour. I can't obviously see any item generation post game start [even spawning monsters use a queue set up at the start] but I think this is good practice. Of course, hacking these tables in the first place is a bad idea and, if we wanted to do that, we should be passing transient copies, but that's a larger refactor.

tmewett commented:
Great work, thank you!
